chisel/src/session.rs

```diff
@@ -0,0 +1,287 @@
+//! ChiselSession
+//!
+//! This module contains the `ChiselSession` struct, which is the top-level
+//! wrapper for a serializable REPL session.
+
+use crate::{prelude::SessionSource, session_source::SessionSourceConfig};
+use ethers_solc::Solc;
+use eyre::Result;
+use foundry_config::SolcReq;
+use serde::{Deserialize, Serialize};
+use std::path::Path;
+use time::{format_description, OffsetDateTime};
+use yansi::Paint;
+
+/// A Chisel REPL Session
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ChiselSession {
+    /// The `SessionSource` object that houses the REPL session.
+    pub session_source: Option<SessionSource>,
+    /// The current session's identifier
+    pub id: Option<String>,
+}
+
+// ChiselSession Common Associated Functions
+impl ChiselSession {
+    /// Create a new `ChiselSession` with a specified `solc` version and configuration.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if there is a failure to install the requested Solc version.
+    pub fn new(config: &SessionSourceConfig) -> Self {
+        // Solc version precidence
+        // - Foundry configuration / `--use` flag
+        // - Latest installed version via SVM
+        // - Default: 0.8.17
+        let solc = Solc::find_or_install_svm_version(
+            if let Some(SolcReq::Version(version)) = config.foundry_config.solc.as_ref() {
+                let version = format!("{}.{}.{}", version.major, version.minor, version.patch);
+                if let Ok(None) = Solc::find_svm_installed_version(&version) {
+                    println!(
+                        "{}",
+                        Paint::green(format!("Installing solidity version {}...", &version))
+                    );
+                }
+                version
+            } else {
+                // If no version was explicitly set, use the latest SVM version.
+                if let Some(version) = Solc::installed_versions().into_iter().max() {
+                    version.to_string()
+                } else {
+                    println!(
+                        "{}",
+                        Paint::green(
+                            "No solidity versions installed! Installing solidity version 0.8.17..."
+                        )
+                    );
+                    String::from("0.8.17")
+                }
+            },
+        );
+
+        // Return initialized ChiselSession with set solc version
+        if let Ok(solc) = solc {
+            Self { session_source: Some(SessionSource::new(&solc, config)), id: None }
+        } else {
+            panic!("Failed to install solidity via svm!");
```

Comment:
would prefer an error here so the function should return a result instead

Reply:
Fixed